controllers/datafoundation.go

```diff
@@ -271,12 +273,21 @@ func (r *dataFoundationReconciler) reconcileStorageCluster() error {
 			return err
 		}
 
-		// Convert the desired size to the device set count based on the underlaying OSD size
-		desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB) / templates.OSDSizeInTiB))
+		// get the desired capacity in TiB planned to be consumed by user in range 1-4TiB
+		desiredOSDSizeInTiB := math.Min(math.Ceil(float64(r.spec.UsableCapacityInTiB)), templates.VerticalScalerUpperBoundInTiB)
 
-		// Get the storage device set count of the current storage cluster
+		// Convert the desired size to the device set count based on the underlying OSD size
+		desiredDeviceSetCount := int(math.Ceil(float64(r.spec.UsableCapacityInTiB)) / desiredOSDSizeInTiB)
+		// Get the storage device set count and OSD size of the current storage cluster
 		currDeviceSetCount := 0
+		currOSDSizeInTiB := 0.0
 		if desiredStorageDeviceSet := findStorageDeviceSet(r.storageCluster.Spec.StorageDeviceSets, defaultDeviceSetName); desiredStorageDeviceSet != nil {
+			// Avoid overflow
+			if len(strconv.Itoa(int(desiredStorageDeviceSet.DataPVCTemplate.Spec.Resources.Requests.Storage().Value()))) <= int(resource.Tera) {
+				return fmt.Errorf("calculating the current OSD size in TiB might result in overflow")
+			}
+			// Get rounded up value and convert to actual
+			currOSDSizeInTiB = float64(desiredStorageDeviceSet.DataPVCTemplate.Spec.Resources.Requests.Storage().ScaledValue(resource.Tera) - 1)
```

Comment:
There is no need to convert this into an int or a float. I believe that resource quantity provides methods so they can be compared.

```suggestion
currOSDSizeInTiB := desiredStorageDeviceSet.DataPVCTemplate.Spec.Resources.Requests.Storage()
```